AppWidget/app/src/main/java/com/example/android/appwidget/glance/image/ImageWorker.kt

```diff
@@ -0,0 +1,160 @@
+            )
+
+            // Temporary workaround to avoid WM provider to disable itself and trigger an
+            // app widget update
+            manager.enqueueUniqueWork(
+                "$uniqueWorkName-workaround",
+                ExistingWorkPolicy.KEEP,
+                OneTimeWorkRequestBuilder<ImageWorker>().apply {
+                    setInitialDelay(Duration.ofDays(365))
+                }.build()
+            )
+        }
+
+        /**
+         * Cancel any ongoing worker
+         */
+        fun cancel(context: Context, glanceId: GlanceId) {
+            WorkManager.getInstance(context).cancelAllWorkByTag(glanceId.toString())
+        }
+    }
+
+    override suspend fun doWork(): Result {
+        return try {
+            val width = inputData.getFloat("width", 0f)
+            val height = inputData.getFloat("height", 0f)
+            val force = inputData.getBoolean("force", false)
+            val uri = getRandomImage(width, height, force)
+            updateImageWidget(width, height, uri)
+            Result.success()
+        } catch (e: Exception) {
+            Log.e(uniqueWorkName, "Error while loading image", e)
+            if (runAttemptCount < 10) {
+                // Exponential backoff strategy will avoid the request to repeat
+                // too fast in case of failures.
+                Result.retry()
+            } else {
+                Result.failure()
+            }
+        }
+    }
+
+    private suspend fun updateImageWidget(width: Float, height: Float, uri: String) {
+        val manager = GlanceAppWidgetManager(context)
+        val glanceIds = manager.getGlanceIds(ImageGlanceWidget::class.java)
+        glanceIds.forEach { glanceId ->
+            updateAppWidgetState(context, glanceId) { prefs ->
+                prefs[ImageGlanceWidget.getImageKey(width, height)] = uri
+                prefs[ImageGlanceWidget.sourceKey] = "Picsum Photos"
+                prefs[ImageGlanceWidget.sourceUrlKey] = "https://picsum.photos/"
+            }
+        }
+        ImageGlanceWidget().updateAll(context)
+    }
+
+    /**
+     * Using Coil and picsum.photos to randomly load images into the cache based on the provided
+     * size. This method returns the path of the cached image so it can be send to the widget.
+     */
+    @OptIn(ExperimentalCoilApi::class)
+    private suspend fun getRandomImage(width: Float, height: Float, force: Boolean): String {
+        val url = "https://picsum.photos/${width.roundToInt()}/${height.roundToInt()}"
+        val request = ImageRequest.Builder(context)
+            .data(url)
+            .build()
+
+        // Request the image to be loaded and throw error if it failed
+        with(context.imageLoader) {
+            if (force) {
+                diskCache?.remove(url)
+                memoryCache?.remove(MemoryCache.Key(url))
+            }
+            val result = execute(request)
+            if (result is ErrorResult) {
```

Comment:
If you extracted this to a helper, then throwing on ErrorResult isn't probably the right option. ImageRequest supports error images, so maybe best to check if the response.drawable is non null.

Reply:
I would need to implement a more complex "state" for the widget with success/error images then. Instead here I decided to throw the error so the worker fails and retries using exponential backoff. 

But agree that probably could be extracted into a more reusable method. But for the sake of the sample I think it's enough

Reply:
Coil has placeholder images also, so a proper library might even use that depending on the delay fetching images.